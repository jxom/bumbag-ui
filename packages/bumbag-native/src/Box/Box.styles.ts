import { styled } from '../styled';
import { theme } from '../utils/theme';

export const StyledBox = styled.View`
  ${theme('native.Box', 'styles.base')};
`;

export const StyledBoxSafe = styled.SafeAreaView`
  ${theme('native.Box.Safe', 'styles.base')};
`;

export const StyledBoxScroll = styled.ScrollView`
  ${theme('native.Box.Scroll', 'styles.base')};
`;

export const StyledBoxKeyboardAvoiding = styled.KeyboardAvoidingView`
  ${theme('native.Box.KeyboardAvoiding', 'styles.base')};
`;

export const StyledBoxTouchable = styled.TouchableOpacity`
  ${theme('native.Box.Touchable', 'styles.base')};
`;

export const StyledBoxTouchableWithoutFeedback = styled.TouchableWithoutFeedback`
  ${theme('native.Box.TouchableWithoutFeedback', 'styles.base')};
`;
