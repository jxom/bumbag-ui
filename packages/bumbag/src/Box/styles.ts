import { css, cssClass } from '../styled';
import { altitude, theme } from '../utils/theme';

const FLEX_HORIZONTAL_ALIGN_MAP = {
  left: 'flex-start',
  center: 'center',
  right: 'flex-end',
};

const FLEX_VERTICAL_ALIGN_MAP = {
  top: 'flex-start',
  center: 'center',
  bottom: 'flex-end',
};

export const style = (styleProps) => cssClass`
  & {
    ${styleProps.style};
  }

  ${getAlignmentAttributes(styleProps)};

  ${
    styleProps.altitude &&
    css`
      & {
        ${altitude(styleProps.altitude)(styleProps)};
      }
    `
  };
`;

export const Box = (styleProps) => cssClass`
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;

  &:focus:not(:focus-visible) {
    outline: none;
  }

  & {
    ${theme(styleProps.themeKey, 'css.root')(styleProps)};
  }
`;

export function getAlignmentAttributes(styleProps) {
  return css`
    ${(styleProps.alignY || styleProps.alignX) &&
    css`
      display: flex;

      ${
        !styleProps.display &&
        css`
          flex-direction: column;
        `
      }

      ${
        styleProps.alignY &&
        css`
          justify-content: ${FLEX_VERTICAL_ALIGN_MAP[styleProps.alignY]};
        `
      }

      ${
        styleProps.alignX &&
        css`
          align-items: ${FLEX_HORIZONTAL_ALIGN_MAP[styleProps.alignX]};
        `
      }
    `}
  `;
}
