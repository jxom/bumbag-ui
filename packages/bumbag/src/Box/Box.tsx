import * as React from 'react';
import { Box as ReakitBox, BoxProps as ReakitBoxProps } from 'reakit';

import { ThemeConfig, CSSProperties } from '../types';
import {
  useStyle,
  useClassName,
  omitCSSProps,
  pickHTMLProps,
  mergeRefs,
  createComponent,
  createElement,
  createHook,
} from '../utils';

import * as styles from './styles';

type ComponentType<R> = React.ComponentType<R> & { useProps: any };

export type LocalBoxProps = {
  use?: string | ComponentType<any>;
  className?: string;
  children?: React.ReactNode | ((props: BoxProps) => React.ReactNode);
  alignX?: 'left' | 'center' | 'right';
  alignY?: 'top' | 'center' | 'bottom';
  altitude?: string;
  variant?: string;
  colorMode?: string;
  /* Component-level theme overrides [Read more](/theming/#component-theming) */
  overrides?: ThemeConfig;
  elementRef?: React.Ref<any>;
  themeKey?: string;
};
export type BoxProps = ReakitBoxProps & CSSProperties & LocalBoxProps;

const useProps = createHook<BoxProps>(
  (_props, { themeKey, themeKeyOverride }) => {
    let props = _props;
    const { use } = props;

    if (use && typeof use !== 'string' && use.useProps) {
      const newProps = use.useProps({ ...props, use: undefined });
      props = { ...props, ...newProps };
    }

    // Convert CSS props to an object.
    // Example input:
    // props = { color: 'red', backgroundColor: 'blue', isEnabled: true }
    //
    // Example output:
    // style = { color: 'red', backgroundColor: 'blue' }
    const style = useStyle(props);

    // Append the styles from above as a className on the DOM element (with precedence).
    let className = useClassName({
      style: styles.style,
      styleProps: { ...props, style },
      themeKey,
      themeKeyOverride,
      prevClassName: props.className,
    });

    // Append the Box styles as a className on the DOM element.
    className = useClassName({
      style: styles.Box,
      styleProps: props,
      prevClassName: className,
      themeKey,
      themeKeyOverride,
    });

    // Pick out and invalid HTML props & omit the CSS props.
    const htmlProps = omitCSSProps(pickHTMLProps({ ...props, className }));

    if (props.elementRef) {
      // @ts-ignore
      htmlProps.ref = mergeRefs(props.elementRef, props.ref);
    }

    if (props.wrapElement) {
      // @ts-ignore
      htmlProps.wrapElement = props.wrapElement;
    }
    return { ...htmlProps };
  },
  { themeKey: 'Box' }
);

export const Box = createComponent<BoxProps>(
  (props) => {
    const boxProps = useProps(props);
    return createElement({ children: props.children, component: ReakitBox, use: props.use, htmlProps: boxProps });
  },
  {
    attach: { displayName: 'Box', useProps },
    themeKey: 'Box',
  }
);
