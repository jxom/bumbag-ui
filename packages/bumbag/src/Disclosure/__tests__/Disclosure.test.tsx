import * as React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { Box } from '../../Box';
import { Disclosure } from '../index';
import render from '../../utils/_tests/render';

describe('props', () => {
  it('should render correctly', () => {
    function Component() {
      const hidden = Disclosure.useState({ baseId: 'test' });
      return <Disclosure {...hidden}>Toggle</Disclosure>;
    }
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should assign a ref', () => {
    const ref = React.createRef();
    function Component() {
      const hidden = Disclosure.useState({ baseId: 'test' });
      return (
        <Disclosure ref={ref} {...hidden}>
          Toggle
        </Disclosure>
      );
    }
    render(<Component />);
    expect(ref.current).toMatchSnapshot();
  });

  it('should render correctly when disabled is set', () => {
    function Component() {
      const hidden = Disclosure.useState({ baseId: 'test' });
      return (
        <Disclosure disabled {...hidden}>
          Toggle
        </Disclosure>
      );
    }
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly when focusable is set', () => {
    function Component() {
      const hidden = Disclosure.useState({ baseId: 'test' });
      return (
        <Disclosure disabled focusable {...hidden}>
          Toggle
        </Disclosure>
      );
    }
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('composition', () => {
  describe('as', () => {
    it('should render correctly', () => {
      function Component() {
        const hidden = Disclosure.useState({ baseId: 'test' });
        return (
          <Disclosure use={Box} {...hidden}>
            Toggle
          </Disclosure>
        );
      }
      const { container } = render(<Component />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hook', () => {
    it('should return with Disclosure props', () => {
      const { result } = renderHook(() => {
        const hidden = Disclosure.useState({ baseId: 'test' });
        return Disclosure.useProps(hidden);
      });
      expect(result.current).toMatchSnapshot();
    });
  });

  describe('render props', () => {
    it('should render correctly', () => {
      function Component() {
        const hidden = Disclosure.useState({ baseId: 'test' });
        return (
          <Disclosure {...hidden}>
            {(DisclosureDisclosureProps) => <div {...DisclosureDisclosureProps}>Hello world</div>}
          </Disclosure>
        );
      }
      const { container } = render(<Component />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});

describe('overrides', () => {
  it('Disclosure.base should render correctly', () => {
    function Component() {
      const hidden = Disclosure.useState({ baseId: 'test' });
      return (
        <Disclosure overrides={{ Disclosure: { styles: { base: { backgroundColor: 'red' } } } }} {...hidden}>
          Toggle
        </Disclosure>
      );
    }
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('theming', () => {
  it('Disclosure.base should render correctly', () => {
    function Component() {
      const hidden = Disclosure.useState({ baseId: 'test' });
      return <Disclosure {...hidden}>Toggle</Disclosure>;
    }
    const { container } = render(<Component />, {
      theme: { Disclosure: { styles: { base: { backgroundColor: 'red' } } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('defaultProps', () => {
  it('should render correctly for className', () => {
    function Component() {
      const hidden = Disclosure.useState({ baseId: 'test' });
      return <Disclosure {...hidden}>Toggle</Disclosure>;
    }
    const { container } = render(<Component />, {
      theme: { Disclosure: { defaultProps: { className: 'red' } } },
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
