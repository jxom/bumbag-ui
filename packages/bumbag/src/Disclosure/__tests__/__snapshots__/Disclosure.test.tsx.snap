// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`composition as should render correctly 1`] = `
.emotion-0 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-0:focus:not(:focus-visible) {
  outline: none;
}

<div
  aria-controls="test"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  role="button"
  tabindex="0"
>
  Toggle
</div>
`;

exports[`composition hook should return with Disclosure props 1`] = `
Object {
  "aria-controls": "test",
  "aria-expanded": false,
  "className": "css-fvskbx css-4lfl7m css-1yuzjtp fp-Box fp-Disclosure",
  "onClick": [Function],
  "onKeyDown": [Function],
  "onKeyUp": [Function],
  "onMouseDown": [Function],
  "ref": [Function],
  "type": "button",
}
`;

exports[`composition render props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<div
  aria-controls="test"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  role="button"
  tabindex="0"
>
  Hello world
</div>
`;

exports[`defaultProps should render correctly for className 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<button
  aria-controls="test"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 red fp-Box fp-Disclosure"
  type="button"
>
  Toggle
</button>
`;

exports[`overrides Disclosure.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0.emotion-0 {
  background-color: #ff0000;
}

<button
  aria-controls="test"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  type="button"
>
  Toggle
</button>
`;

exports[`props should assign a ref 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<button
  aria-controls="test"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  type="button"
>
  Toggle
</button>
`;

exports[`props should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<button
  aria-controls="test"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  type="button"
>
  Toggle
</button>
`;

exports[`props should render correctly when disabled is set 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<button
  aria-controls="test"
  aria-disabled="true"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  disabled=""
  style="pointer-events: none;"
  type="button"
>
  Toggle
</button>
`;

exports[`props should render correctly when focusable is set 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

<button
  aria-controls="test"
  aria-disabled="true"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  style="pointer-events: none;"
  type="button"
>
  Toggle
</button>
`;

exports[`theming Disclosure.base should render correctly 1`] = `
.emotion-1 {
  margin: unset;
  padding: unset;
  border: unset;
  background: unset;
  font: unset;
  font-family: inherit;
  font-size: 100%;
  box-sizing: border-box;
}

.emotion-1:focus:not(:focus-visible) {
  outline: none;
}

.emotion-0.emotion-0 {
  background-color: #ff0000;
}

<button
  aria-controls="test"
  aria-expanded="false"
  class="emotion-0 emotion-1 emotion-2 fp-Box fp-Disclosure"
  type="button"
>
  Toggle
</button>
`;
